packages/nimble/R/BUGS_macros.R

```diff
@@ -0,0 +1,126 @@
+#'
+#' @param unpackArgs (TRUE or FALSE) Should arguments be passed as a list (FALSE) or as separate arguments (TRUE)?  See details and examples.
+#'
+#' @details The arguments \code{use3pieces} and \code{unpackArgs} indicate how \code{fun} expects to have arguments arranged from an input line of code (processed by \code{nimbleModel}).
+#'
+#' Consider the defaults \code{use3pieces = TRUE} and \code{unpackArgs = TRUE}, for a macro called \code{macro1}.  In this case, the line of model code \code{x ~ macro1(arg1 = z[1:10], arg2 = "hello")} will be passed to \code{fun} as \code{fun(stoch = TRUE, LHS = x, arg1 = z[1:10], arg2 = "hello")}.
+#'
+#' If \code{use3pieces = TRUE} but \code{unpackArgs = FALSE}, then the RHS will be passed as is, without unpacking its arguments into separate arguments to \code{fun}.  In this case, \code{x ~ macro1(arg1 = z[1:10], arg2 = "hello")} will be passed to \code{fun} as \code{fun(stoch = TRUE, LHS = x, RHS = macro1(arg1 = z[1:10], arg2 = "hello"))}.
+#'
+#' If \code{use3pieces = FALSE} and \code{unpackArgs = FALSE}, the entire line of code is passed as a single object.  In this case, \code{x ~ macro1(arg1 = z[1:10], arg2 = "hello")} will be passed to \code{fun} as \code{fun(x ~ macro1(arg1 = z[1:10], arg2 = "hello"))}.  It is also possible in this case to pass a macro without using a \code{~} or \code{<-}.  For example, the line \code{macro1(arg1 = z[1:10], arg2 = "hello")} will be passed to \code{fun} as \code{fun(macro1(arg1 = z[1:10], arg2 = "hello"))}.
+#'
+#' If \code{use3pieces = FALSE} and \code{unpackArgs = TRUE}, it won't make sense to anticipate a declaration using \code{~} or \code{<-}.  Instead, arguments from an arbitrary call will be passed as separate arguments.  For example, the line \code{macro1(arg1 = z[1:10], arg2 = "hello")} will be passed to \code{fun} as \code{fun(arg1 = z[1:10], arg2 = "hello")}.
+#'
+#' It is extremely useful to be familiar with processing R code as an object to write \code{fun} correctly.  Functions such as \code{\link{substitute}} and \code{\link{as.name}} (e.g. \code{as.name('~')}), \code{\link{quote}}, \code{\link{parse}} and \code{\link{deparse}} are particularly handy.
+#'
+#' Multiple lines of new code should be contained in \code{ {} }. Extra curly braces are not a problem. See example 2.
+#'
+#' Macro expansion is done recursively: One macro can return code that invokes another macro.
+#' 
+#' @return A list with a named element \code{code} that contains the replacement code.
+#' 
+#' @examples
+#' nimbleOptions(enableModelMacros = TRUE)
+#' nimbleOptions(verbose = FALSE)
+#' 
+#' ## Example 1: Say one is tired of writing "for" loops.
+#' ## This macro will generate a "for" loop with dnorm declarations
+#' all_dnorm <- model_macro_builder(
+#'     function(stoch, LHS, RHSvar, start, end, sd = 1) {
+#'         newCode <- substitute(
+#'             for(i in START:END) {
```

Comment:
As @paciorek mentions, variable collision is a major problem. In this example, `i` can collide with bound variables, e.g.
```r
for (i in 1:10) {
  x[i,1:10] ~ all_dnorm(mu, start = 1, end = 10)
}
```
(there's probably also an indexing issue)

Some possible solutions include:
- [alpha renaming](https://en.wikipedia.org/wiki/Lambda_calculus#Alpha_equivalence) bound variables to avoid conflict (in this case, renaming the macro's `i` to avoid conflict). I believe this could be implemented within `model_macro_builder` in a three step process: (1) rename all variables in the arguments; (2) expand macro; (3) revert all variable names in arguments and resolve name conflicts, e.g. by adding `_1`, `_2` suffices.
- provide a helper `fresh_variable()` to create a variable with no collisions. This would involve either (1) lots of wiring to pass around the set of variables to avoid, or (2) creating a global fresh variable id (which makes testing difficult).